### PR TITLE
✨ Is3318/resend 2fh handler  (⚠️ devops)  

### LINKED ISSUE #3318: Resend phone confirmation code
Workflow discussed with @odeimaiz 

- During registration/login
  - ``/auth/validate-code-register``  (``auth_validate_2fa_register``) returns a retry identification
  - wait 60 secs
  - user can ask for code resend ``auth_validate_2fa_register_retry(identification)``
     - If ``identification`` does not exist or did not pass enough time, it fails to retry

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

Implements resend 2FA and creates a mechanism to protect "continuation" routes. 
image
### Highlights:

-✨New ``services/webserver`` api version: ``0.12.2 → 0.13.0``
-✨New API entry ``POST auth/two_factor:resend`` with SMS or Email as options
-✨New Response of ``POST auth/login``: 
```json
{
  "data": {
    "name": "string",
    "parameters": {
      "message": "string",
      "retry_2fa_after": 60,
      "next_url": "string"
    }
  },
  "error": null
}
```
-✨New ``session_access`` decorators for access control
   - Many of the login workflows requires more that one call to the API: e.g. login -> (resend SMS? x times) -> login_2fa. 
   - All these routes do not require authentication and therefore exposed in the open
   - We use this technique to protect the entrypoint 
   - The API will reject access if the route is not part of a pre-defined workflow (e.g. cannot resend SMS w/o having called login) 
- ♻️ Extra cleanup of handlers and tests
     - improved ``api/specs/webserver/scripts`` 
     - regenerated ``api/specs/webserver/openapi-auth.yaml``
-  🔨 ``testit`` marker: marks test to run during development. SEE ``launch.template.json``
- 🐼  cleanup email templates

##  ⚠️ devops

- New variable ``LOGIN_2FA_CODE_EXPIRATION_SEC`` defaults to 60

## Related issue/s

- closes #3318 

## How to test

- Manually checked with @odeimaiz 
- Testsuite
```cmd
cd services/web/server
make install-dev
pytest tests/unit/with_db/03/login
```

## Checklist
- [x] ``make version-*``
- [x] ``make openapi.json``
- [x] Unit tests for the changes exist
- [x] Runs in the swarm

### CODE REVIEW

**Review by odeimaiz — APPROVED** [review]
Thanks!!

File: .vscode/launch.template.json [hunk 1]
```diff
@@ -23,6 +23,25 @@
...
+      "name": "Python: Testit",
+      "type": "python",
+      "request": "launch",
+      "module": "pytest",
+      "args": [
+        "--ff",
+        "--log-cli-level=INFO",
+        "--pdb",
+        "--setup-show",
+        "-m testit",
```
Comment (sanderegg) [inline .vscode/launch.template.json:36]: cool

File: packages/pytest-simcore/src/pytest_simcore/helpers/utils_envs.py [hunk 2]
```diff
@@ -17,7 +17,7 @@
 #
 # monkeypatch using dict
 #
-def setenvs_from_dict(monkeypatch: MonkeyPatch, envs: EnvVarsDict) -> EnvVarsDict:
+def setenvs_from_dict(monkeypatch: MonkeyPatch, envs: dict[str, Any]) -> EnvVarsDict:
```
Comment (sanderegg) [inline packages/pytest-simcore/src/pytest_simcore/helpers/utils_envs.py:20]: not sure I get why you changed this?

File: services/web/server/src/simcore_service_webserver/login/_2fa.py [hunk 3]
```diff
@@ -108,14 +126,18 @@ def _sender():
...
+    await asyncio.get_event_loop().run_in_executor(executor=None, func=_sender)
 
 
 #
 # EMAIL
 #
 
 
+class EmailError(RuntimeError):
+    pass
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/login/_2fa.py:138]: would it make sense to have an ```errors.py``` module in the login folder?

File: services/web/server/src/simcore_service_webserver/login/handlers_registration.py [hunk 4]
```diff
@@ -186,6 +188,9 @@ class RegisterPhoneBody(InputSchema):
 
 
 @global_rate_limit_route(number_of_requests=5, interval_seconds=MINUTE)
+@session_access_constraint(
+    allow_access_after=["auth_register", "auth_login"], max_number_of_access=1
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/login/handlers_registration.py:192]: very nice!

File: services/web/server/tests/unit/isolated/test_templates.py [hunk 5]
```diff
@@ -33,7 +33,7 @@ def test_all_email_templates_include_subject(template_path: Path, app: web.Appli
     subject, content = template_path.read_text().split("\n", 1)
 
     assert re.match(
-        r"[\{\}a-zA-Z0-9\-_\s]+", subject
+        r"[\{\}a-zA-Z0-9\-_\s]+", subject.strip("🐼")
```
Comment (sanderegg) [inline services/web/server/tests/unit/isolated/test_templates.py:36]: panda?

**Review by sanderegg — APPROVED** [review]
partially pair reviewed. great work!

File: services/web/server/tests/unit/with_dbs/03/login/test_login_registration.py [hunk 6]
```diff
@@ -133,18 +134,20 @@ async def test_registration_with_expired_confirmation(
         autospec=True,
         return_value=LoginSettings(
             LOGIN_REGISTRATION_CONFIRMATION_REQUIRED=True,
-            LOGIN_REGISTRATION_INVITATION_REQUIRED=True,
+            LOGIN_REGISTRATION_INVITATION_REQUIRED=True,  # <----- invitation REQUIRED
```
Comment (GitHK) [inline services/web/server/tests/unit/with_dbs/03/login/test_login_registration.py:137]: The comment tells me nothing more than what I already can read. Could you maybe change it or remove it?

File: services/web/server/tests/unit/with_dbs/03/login/test_login_registration.py [hunk 7]
```diff
@@ -170,26 +173,24 @@ async def test_registration_with_invalid_confirmation_code(
         autospec=True,
         return_value=LoginSettings(
             LOGIN_REGISTRATION_CONFIRMATION_REQUIRED=True,
-            LOGIN_REGISTRATION_INVITATION_REQUIRED=False,
+            LOGIN_REGISTRATION_INVITATION_REQUIRED=False,  # <----- NO invitation
```
Comment (GitHK) [inline services/web/server/tests/unit/with_dbs/03/login/test_login_registration.py:176]: I'd say the same here

**Review by GitHK — APPROVED** [review]
looks nice

File: services/web/server/tests/unit/with_dbs/03/login/test_login_registration.py (hunk 6 above)
Comment (pcrespov) [reply services/web/server/tests/unit/with_dbs/03/login/test_login_registration.py:137]: They highlight a the variable that was change with respect to the previous test configuration

File: services/web/server/src/simcore_service_webserver/login/_2fa.py (hunk 3 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/login/_2fa.py:138]: Will do in the next PR where I will actulally be using these errors.

**Pull request merged**